mlir/unittests/Target/LLVM/SerializeROCDLTarget.cpp

```diff
@@ -158,3 +158,69 @@ TEST_F(MLIRTargetLLVMROCDL, SKIP_WITHOUT_AMDGPU(SerializeROCDLToBinary)) {
     ASSERT_FALSE(object->empty());
   }
 }
+
+// Test ROCDL metadata.
+TEST_F(MLIRTargetLLVMROCDL, SKIP_WITHOUT_AMDGPU(GetELFMetadata)) {
+  if (!hasROCMTools())
+    GTEST_SKIP() << "ROCm installation not found, skipping test.";
```

Comment:
What are we depending on the ROCm installation for exactly? The only hard dependency is HSA in `offload/` at least. The ROCm Device Libs basically just provide math if you're using the appropriate LLVM builtins.

Unrelated, but I've always wondered if we could just make the math dependencies internal by relying on what's in `libclc` or LLVM libc's math (The latter is missing a few functions like hyperbolic f64 trig and gamma functions).

Reply:
In this test, we go from MLIR to binary and then check the metadata.
Now, for going from MLIR to object we currently need to find the `ld.lld` from the ROCm toolchain, see:  https://github.com/llvm/llvm-project/blob/main/mlir/lib/Target/LLVM/ROCDL/Target.cpp#L380-L384

Replacing all the math deps with in-tree alternatives would be cool. However, I don't know if vendors would like that.

Reply:
Seems weird that we need `ld.lld` of all things, as far as I'm aware the ROCm one is the exact same as the one you can build in-tree.

Reply:
We actually use the in-tree one downstream in rocMLIR

What we really need is the ability to go from a `.o` to a `.so` - it'd be nice if that were a library function that could be called instead of needing to invoke `ld`

(and if there's target-specific logic about creating that `.so` - the HSACO -> "HSA Code Object" - then we'd need that)

Reply:
I remember discussing that with @MaskRay, but I remember he said that the `ld.lld` interface did poorly with reentrancy so he didn't want to export it. I think that it would be really useful for our case given that NVIDIA has https://docs.nvidia.com/cuda/ptx-compiler-api/index.html to create usable objects within the same process.

Reply:
The one in tree also works, for example, in rocMLIR we build the in-tree version https://github.com/ROCm/rocMLIR/blob/develop/mlir/lib/Target/Target.cpp#L183-L188 .

However, this implementation is meant to be shipped only with MLIR, and since we might need device libraries the best option is depending in ROCm.